### PR TITLE
add rpm build target

### PR DESCRIPTION
**Pull request by Andor**
based on https://github.com/utobi/prometheus-rpm

### CODE REVIEW

File: prometheus.spec [hunk 1]
```diff
@@ -0,0 +1,99 @@
...
+
+Name:           prometheus
+Version:        %{version}
+Release:        1%{?dist}
+Summary:        Prometheus is a systems and service monitoring system. It collects metrics from configured targets at given intervals, evaluates rule expressions, displays the results, and can trigger alerts if some condition is observed to be true.
+Group:          System Environment/Daemons
+License:        See the LICENSE file at github.
+URL:            https://github.com/prometheus/prometheus
+Requires(pre):  /usr/sbin/useradd
+Requires:       systemd-units
```
Comment (brian-brazil) [inline prometheus.spec]: Which distro is this for? Systemd is not widespread enough to presume it's everywhere, and in the deb world anyway it'll be at least 2019 until that's the case. I imagine it's longer for the rpm world given RHEL.

**Comment by brian-brazil** [comment]
If we're providing RPMs we need them to work on all RPM-based distributions that are still supported or in common use. We may also want to have any packaging like this done centrally, rather than on a repo by repo basis as we do have rather a few repos.

@sdurrheimer

**Comment by fabxc** [comment]
Interesting, thanks.

We recently started a discussion around build and release tooling.
Maybe this could be fitted into that, so we can have rpm packages for other components, too.

https://docs.google.com/document/d/1Ql-f_aThl-2eB5v3QdKV_zgBdetLLbdxxChpy-TnWSE/edit#heading=h.24x0bg1hyuak

@sdurrheimer 
I don't want to overload and delay the initial version too much. Just something to think about.

**Comment by Andor** [comment]
Thanks for the feedback, I will improve this MR soon
I will add support for no-systemd distros: fedora<21, centos/redhat<7, suse<1210.

**Comment by sdurrheimer** [comment]
Given that we are going on the way of externalizing the building/packaging process of our components, I'm not really for putting this in the project repo, but more in a separate repo and it should also be automated.

@fabc, it can be interesting to put that on the release process, but then we should also provide `.deb` packages. Also, I'm a little bit worried about putting more and more thing in a single CircleCI build.

File: Makefile [hunk 2]
```diff
@@ -48,11 +50,22 @@ tarballs:
...
+	@echo ">> building rpm package for no-systemd distros"
+	@rpmbuild --buildroot "$(PWD)/.build/nosystemd" -D "_topdir $(PWD)" -D "_builddir $(PWD)/.build/nosystemd" -D "src_root $(PWD)" -D "version $(VERSION)" -D 'use_systemd 0' -bb prometheus.spec
+	@mv RPMS/*/*.rpm "$(PWD)"/
+
+rpm-systemd:
+	@echo ">> building rpm package for systemd distros"
+	@rpmbuild --buildroot "$(PWD)/.build/systemd" -D "_topdir $(PWD)" -D "_builddir $(PWD)/.build/systemd" -D "src_root $(PWD)" -D "version $(VERSION)" -D 'use_systemd 1' -bb prometheus.spec
+	@mv RPMS/*/*.rpm "$(PWD)"/
+
+rpm:	build rpm-systemd rpm-nosystemd
```
Comment (brian-brazil) [inline Makefile]: Can you not produce a single RPM? I imagine there's more dimensions that just systemd and a combinatorial explosion is unwise. It'd be best to only provide a standard init script, as that's going to work everywhere.

File: prometheus.spec [hunk 3]
```diff
@@ -0,0 +1,115 @@
...
+install -m 755 "%{src_root}/promtool" usr/bin/promtool
+
+%if 0%{use_systemd}
+install -d lib/systemd/system
+install -m 644 "%{src_root}/rpm/prometheus.service" lib/systemd/system/prometheus.service
+%else
+install -d etc/init.d
+install -m 755 "%{src_root}/rpm/prometheus.init" etc/init.d/prometheus
+install -d etc/logrotate.d
+install -m 644 "%{src_root}/rpm/prometheus.logrotate" etc/logrotate.d/prometheus
```
Comment (brian-brazil) [inline prometheus.spec]: This RPM as configured produces no logs.

File: prometheus.spec [hunk 4]
```diff
@@ -0,0 +1,115 @@
...
+%pre
+getent group prometheus >/dev/null || groupadd -r prometheus
+getent passwd prometheus >/dev/null || \
+  useradd -r -g prometheus -s /sbin/nologin \
+    -d /var/lib/prometheus -c prometheus prometheus
+exit 0
+
+%post
+chgrp prometheus /var/run/prometheus
+chmod 774 /var/run/prometheus
```
Comment (brian-brazil) [inline prometheus.spec]: Why 774 rather than 775? r is useless on a directory without x.

Comment (Andor) [inline prometheus.spec]: Oops, copy-pasted from the original project.

File: Makefile (hunk 2 above)
Comment (Andor) [inline Makefile]: We include binary in package with ldd depends only from libc:
```
linux-vdso.so.1 =>  (0x00007ffd2429d000)
libpthread.so.0 => /lib/x86_64-linux-gnu/libpthread.so.0 (0x00007f75ffcde000)
libc.so.6 => /lib/x86_64-linux-gnu/libc.so.6 (0x00007f75ff914000)
/lib64/ld-linux-x86-64.so.2 (0x000055a6eaafa000)
```


So we have almost nothing in dependencies but init-system, and we can create universal package for all systemd distros and another one for all nosystemd distros.

File: prometheus.spec (hunk 3 above)
Comment (Andor) [inline prometheus.spec]: No, if we are not using systemd, then we install SysV init-script:
```
$ grep LOG_FILE rpm/prometheus.init 
LOG_FILE=/var/log/prometheus/prometheus.log
  daemonize -u ${DAEMON_USER} -p ${PID_FILE} -l ${LOCK_FILE} -a -e ${LOG_FILE} -o ${LOG_FILE} ${DAEMON} ${ARGS}
```
And if we are using systemd, then logs managed by systemd.

**Comment by d-ulyanov** [comment]
+1

**Comment by Andor** [comment]
Merge request updated, we now able to create i386 and amd64 packages for systemd and for non-systemd.

**Comment by Andor** [comment]
And I will add Travis tests for these rpm's soon.

**Pull request closed without merging**